File: Clojure/problem003.clj
#! /usr/bin/env clojure
; Problem 3 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=3
(require '[mathtools :refer [primes]])

(defn maxprimefactor' [x ps]
  (if (empty? ps)
    x
    (if (zero? (mod x (first ps)))
      (let [qx (quot x (first ps))]
        (if (= qx 1)
          x
          (recur qx ps)))
      (recur x (rest ps)))))

(defn maxprimefactor [x]
  (int (maxprimefactor' x (primes (Math/sqrt x)))))

(println (maxprimefactor 600851475143N))

File: Python/problem003.py
#! /usr/bin/env python3
'''
Problem 3 - Project Euler
http://projecteuler.net/index.php?section=problems&id=003
'''
import math
from mathtools import primes

def maxprimefactor(x):
    p = 2
    while x != 1:
        if x % p == 0:
            x = x // p
        else:
            p += 2 if p != 2 else 1
    return p

if __name__ == '__main__':
    N = 600851475143
    assert(maxprimefactor(2) == 2)
    assert(maxprimefactor(3) == 3)
    assert(maxprimefactor(5) == 5)
    assert(maxprimefactor(10) == 5)
    print(maxprimefactor(N))
